src/sentry/uptime/grouptype.py

```diff
@@ -1,15 +1,289 @@
     ProjectUptimeSubscriptionMode,
 )
-from sentry.workflow_engine.types import DetectorSettings
+from sentry.utils import metrics
+from sentry.workflow_engine.handlers.detector.base import DetectorOccurrence, EventData
+from sentry.workflow_engine.handlers.detector.stateful import (
+    DetectorThresholds,
+    StatefulDetectorHandler,
+)
+from sentry.workflow_engine.models import DataPacket, Detector
+from sentry.workflow_engine.processors.data_condition_group import ProcessedDataConditionGroup
+from sentry.workflow_engine.types import (
+    DetectorEvaluationResult,
+    DetectorGroupKey,
+    DetectorPriorityLevel,
+    DetectorSettings,
+)
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class UptimePacketValue:
+    """
+    Represents the value passed into the uptime detector
+    """
+
+    check_result: CheckResult
+    subscription: UptimeSubscription
+    metric_tags: dict[str, str]
+
+
+def build_detector_fingerprint_component(detector: Detector) -> str:
+    return f"uptime-detector:{detector.id}"
+
+
+def build_fingerprint(detector: Detector) -> list[str]:
+    return [build_detector_fingerprint_component(detector)]
+
+
+def get_active_failure_threshold() -> int:
+    """
+    When in active monitoring mode, overrides how many failures in a row we
+    need to see to mark the monitor as down
+    """
+    return options.get("uptime.active-failure-threshold")
+
+
+def get_active_recovery_threshold() -> int:
+    """
+    When in active monitoring mode, how many successes in a row do we need to
+    mark it as up
+    """
+    return options.get("uptime.active-recovery-threshold")
+
+
+def build_evidence_display(result: CheckResult) -> list[IssueEvidence]:
+    evidence_display: list[IssueEvidence] = []
+
+    status_reason = result["status_reason"]
+    if status_reason:
+        reason_evidence = IssueEvidence(
+            name="Failure reason",
+            value=f'{status_reason["type"]} - {status_reason["description"]}',
+            important=True,
+        )
+        evidence_display.extend([reason_evidence])
+
+    duration_evidence = IssueEvidence(
+        name="Duration",
+        value=f"{result["duration_ms"]}ms",
+        important=False,
+    )
+    evidence_display.append(duration_evidence)
+
+    request_info = result["request_info"]
+    if request_info:
+        method_evidence = IssueEvidence(
+            name="Method",
+            value=request_info["request_type"],
+            important=False,
+        )
+        status_code_evidence = IssueEvidence(
+            name="Status Code",
+            value=str(request_info["http_status_code"]),
+            important=False,
+        )
+        evidence_display.extend([method_evidence, status_code_evidence])
+
+    return evidence_display
+
+
+def build_event_data(result: CheckResult, detector: Detector) -> EventData:
+    # Default environment when it hasn't been configured
+    env = detector.config.get("environment", "prod")
```

Comment:
is this normal behavior? Does every detector need this? Should we at least make the default a constant?

Reply:
This is specific to the uptime detector. Unfortunately we hardcode default environment names a number of places in the sentry backend, so this is par for the course right now